pkg/eventpersisterops/persister/goal_event.go

```diff
@@ -0,0 +1,307 @@
+				zap.String("featureId", tr.featureID),
+				zap.String("environmentNamespace", environmentNamespace),
+			)
+			handledCounter.WithLabelValues(codeFailedToFindFeatureVersion).Inc()
+			return false, ErrFailedToFindFeatureVersion
+		}
+		// Update the user count by rule
+		err := u.updateUserCountByRule(
+			environmentNamespace,
+			tr.featureID,
+			fVersion,
+			event.UserId,
+			tr,
+		)
+		if err != nil {
+			return true, err
+		}
+	}
+	return false, nil
+}
+
+func (u *evalGoalUpdater) listAutoOpsRules(
+	ctx context.Context,
+	environmentNamespace string,
+) ([]*aoproto.AutoOpsRule, error) {
+	exp, err, _ := u.flightgroup.Do(
+		fmt.Sprintf("%s:%s", environmentNamespace, "listAutoOpsRules"),
+		func() (interface{}, error) {
+			aor := []*aoproto.AutoOpsRule{}
+			cursor := ""
+			for {
+				resp, err := u.autoOpsClient.ListAutoOpsRules(ctx, &aoproto.ListAutoOpsRulesRequest{
+					EnvironmentNamespace: environmentNamespace,
+					PageSize:             listRequestSize,
+					Cursor:               cursor,
+				})
+				if err != nil {
+					return nil, err
+				}
+				aor = append(aor, resp.AutoOpsRules...)
+				aorSize := len(resp.AutoOpsRules)
+				if aorSize == 0 || aorSize < listRequestSize {
+					return aor, nil
+				}
+				cursor = resp.Cursor
+			}
+		},
+	)
+	if err != nil {
+		return nil, err
+	}
+	return exp.([]*aoproto.AutoOpsRule), nil
+}
+
+func (u *evalGoalUpdater) findOpsRules(
+	goalID string,
+	listAutoOpsRules []*aoproto.AutoOpsRule,
+) ([]string, []*goalTargetRule) {
+	featureIDsMap := make(map[string]struct{})
+	targetRules := []*goalTargetRule{}
+	for _, aor := range listAutoOpsRules {
+		autoOpsRule := &aodomain.AutoOpsRule{AutoOpsRule: aor}
+		// We ignore the rules that are already triggered
+		if autoOpsRule.AlreadyTriggered() {
+			continue
+		}
+		clauses, err := autoOpsRule.ExtractOpsEventRateClauses()
+		if err != nil {
+			handledCounter.WithLabelValues(codeFailedToExtractOpsEventRateClauses).Inc()
+			continue
+		}
+		if len(clauses) == 0 {
+			continue
+		}
+		// Find the clauses that contain the goal ID from the the goal event
+		targetClauses := make(map[string]*aoproto.OpsEventRateClause)
+		for id, clause := range clauses {
+			if clause.GoalId == goalID {
+				featureIDsMap[autoOpsRule.FeatureId] = struct{}{}
+				targetClauses[id] = clause
+			}
+		}
+		if len(targetClauses) == 0 {
+			continue
+		}
+		targetRules = append(targetRules, &goalTargetRule{
+			id:        autoOpsRule.Id,
+			featureID: autoOpsRule.FeatureId,
+			clauses:   targetClauses,
+		})
+	}
+	// Convert map to slice
+	featureIDs := make([]string, 0, len(featureIDsMap))
+	for id := range featureIDsMap {
+		featureIDs = append(featureIDs, id)
+	}
+	return featureIDs, targetRules
+}
+
+func (u *evalGoalUpdater) findFeatureVersion(
+	featureID string,
+	features []*featureproto.Feature,
+) int32 {
+	for _, f := range features {
+		if f.Id == featureID {
+			return f.Version
+		}
+	}
+	return 0
```

Comment:
FYI: Historically, returning -1 represents an error, so if we want to express that not finding feature version is an error, it's good to return -1.

https://www.quora.com/What-is-the-significance-or-meaning-of-using-return-1-in-C
https://developer.mozilla.org/ja/docs/Web/JavaScript/Reference/Global_Objects/String/indexOf
https://docs.python.org/3/library/stdtypes.html#str.find